core/node/prover_api/src/processor.rs

```diff
@@ -0,0 +1,196 @@
+use std::sync::Arc;
+
+use axum::{extract::Path, Json};
+use zksync_basic_types::{
+    basic_fri_types::Eip4844Blobs, commitment::L1BatchCommitmentMode, L1BatchNumber,
+};
+use zksync_dal::{ConnectionPool, Core, CoreDal};
+use zksync_object_store::{bincode, ObjectStore};
+use zksync_prover_interface::{
+    api::{
+        ProofGenerationData, ProofGenerationDataRequest, ProofGenerationDataResponse,
+        VerifyProofRequest,
+    },
+    inputs::{
+        L1BatchMetadataHashes, VMRunWitnessInputData, WitnessInputData, WitnessInputMerklePaths,
+    },
+    outputs::L1BatchProofForL1,
+};
+
+use crate::error::ProcessorError;
+
+#[derive(Clone)]
+pub(crate) struct Processor {
+    blob_store: Arc<dyn ObjectStore>,
+    pool: ConnectionPool<Core>,
+    commitment_mode: L1BatchCommitmentMode,
+    last_available_batch_number: L1BatchNumber,
+    last_available_batch_data: Option<Box<ProofGenerationData>>,
+}
+
+impl Processor {
+    pub(crate) fn new(
+        blob_store: Arc<dyn ObjectStore>,
+        pool: ConnectionPool<Core>,
+        commitment_mode: L1BatchCommitmentMode,
+        last_available_batch_number: L1BatchNumber,
+    ) -> Self {
+        Self {
+            blob_store,
+            pool,
+            commitment_mode,
+            last_available_batch_number,
+            last_available_batch_data: None,
+        }
+    }
+
+    #[tracing::instrument(skip_all)]
+    pub(crate) async fn get_proof_generation_data(
+        &mut self,
+        request: Json<ProofGenerationDataRequest>,
+    ) -> Result<Json<ProofGenerationDataResponse>, ProcessorError> {
+        tracing::info!("Received request for proof generation data: {:?}", request);
+
+        let l1_batch_number = self
+            .pool
+            .connection()
+            .await
+            .unwrap()
+            .proof_generation_dal()
+            .get_available_batch()
+            .await?
+            .unwrap_or(self.last_available_batch_number);
+
+        if let Some(data) = &self.last_available_batch_data {
+            if data.l1_batch_number == l1_batch_number {
+                return Ok(Json(ProofGenerationDataResponse::Success(Some(
+                    data.clone(),
+                ))));
+            }
+        }
+
+        let proof_generation_data = self
+            .proof_generation_data_for_existing_batch(l1_batch_number)
+            .await;
+
+        match proof_generation_data {
+            Ok(data) => {
+                self.last_available_batch_number = l1_batch_number;
+                self.last_available_batch_data = Some(Box::new(data.clone()));
+                Ok(Json(ProofGenerationDataResponse::Success(Some(Box::new(
+                    data,
+                )))))
+            }
+            Err(err) => Err(err),
+        }
+    }
+
+    #[tracing::instrument(skip(self))]
+    async fn proof_generation_data_for_existing_batch(
```

Comment:
This is copy-pasted from the proof data handler, right? Maybe makes sense to extract this functionality into a function there, add it as a dependency, and reuse it here, just so that we avoid code duplication.